// @nuxt/ui@4.3.0/dist/runtime/components/ChatPalette.d.vue.ts
import type { AppConfig } from '@nuxt/schema';
import theme from '#build/ui/chat-palette';
import type { ComponentConfig } from '../types/tv';
type ChatPalette = ComponentConfig<typeof theme, AppConfig, 'chatPalette'>;
export interface ChatPaletteProps {
    /**
     * The element or component this component should render as.
     * @defaultValue 'div'
     */
    as?: any;
    class?: any;
    ui?: ChatPalette['slots'];
}
export interface ChatPaletteSlots {
    default(props?: {}): any;
    prompt(props?: {}): any;
}
declare const __VLS_export: __VLS_WithSlots<import("vue").DefineComponent<ChatPaletteProps, {}, {}, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, {}, string, import("vue").PublicProps, Readonly<ChatPaletteProps> & Readonly<{}>, {}, {}, {}, {}, string, import("vue").ComponentProvideOptions, false, {}, any>, ChatPaletteSlots>;
declare const _default: typeof __VLS_export;
export default _default;
type __VLS_WithSlots<T, S> = T & {
    new (): {
        $slots: S;
    };
};
