// @nuxt/ui@4.3.0/dist/runtime/components/PageLinks.d.vue.ts
import type { AppConfig } from '@nuxt/schema';
import theme from '#build/ui/page-links';
import type { IconProps, LinkProps } from '../types';
import type { ComponentConfig } from '../types/tv';
type PageLinks = ComponentConfig<typeof theme, AppConfig, 'pageLinks'>;
export interface PageLink extends Omit<LinkProps, 'custom'> {
    label: string;
    /**
     * @IconifyIcon
     */
    icon?: IconProps['name'];
    class?: any;
    ui?: Pick<PageLinks['slots'], 'item' | 'link' | 'linkLabel' | 'linkLabelExternalIcon' | 'linkLeadingIcon'>;
}
export interface PageLinksProps<T extends PageLink = PageLink> {
    /**
     * The element or component this component should render as.
     * @defaultValue 'nav'
     */
    as?: any;
    title?: string;
    links?: T[];
    class?: any;
    ui?: PageLinks['slots'];
}
type SlotProps<T> = (props: {
    link: T;
    active: boolean;
    ui: PageLinks['ui'];
}) => any;
export interface PageLinksSlots<T extends PageLink = PageLink> {
    'title'(props?: {}): any;
    'link': SlotProps<T>;
    'link-leading': SlotProps<T>;
    'link-label'(props: {
        link: T;
        active: boolean;
    }): any;
    'link-trailing'(props: {
        link: T;
        active: boolean;
    }): any;
}
declare const __VLS_export: <T extends PageLink>(__VLS_props: NonNullable<Awaited<typeof __VLS_setup>>["props"], __VLS_ctx?: __VLS_PrettifyLocal<Pick<NonNullable<Awaited<typeof __VLS_setup>>, "attrs" | "emit" | "slots">>, __VLS_expose?: NonNullable<Awaited<typeof __VLS_setup>>["expose"], __VLS_setup?: Promise<{
    props: __VLS_PrettifyLocal<PageLinksProps<T>> & import("vue").PublicProps & (typeof globalThis extends {
        __VLS_PROPS_FALLBACK: infer P;
    } ? P : {});
    expose: (exposed: {}) => void;
    attrs: any;
    slots: PageLinksSlots<T>;
    emit: {};
}>) => import("vue").VNode & {
    __ctx?: Awaited<typeof __VLS_setup>;
};
declare const _default: typeof __VLS_export;
export default _default;
type __VLS_PrettifyLocal<T> = {
    [K in keyof T as K]: T[K];
} & {};
